test/15_target-hostapp.ts

```diff
@@ -340,73 +342,87 @@ export default () => {
 					);
 			});
 
-			it('should succeed in PATCHing device to greater version', async () => {
-				await supertest(admin)
-					.patch(`/${version}/device(${device.id})`)
-					.send({ should_be_operated_by__release: upgradeReleaseId })
-					.expect(200);
-				const { body } = await supertest(admin).get(
-					`/${version}/device(${device.id})?$select=should_be_operated_by__release`,
-				);
-				expect(body.d[0]).to.not.be.undefined;
-				expect(body.d[0]['should_be_operated_by__release'].__id).to.equal(
-					upgradeReleaseId,
-				);
-			});
+			(
+				[
+					[
+						'semver & release_tag',
+						'balenaOS 2.51.0+rev1',
+						() => prodTagAndSemverNucHostappReleaseId,
+					],
+					[
+						'semver only',
+						'balenaOS 2.51.0+rev2',
+						() => prodSemverOnlyNucHostappReleaseId,
+					],
+				] as const
+			).forEach(([titlePart, osVersion, getUpgradeReleaseId]) => {
+				it(`should succeed in PATCHing device to a greater version (with ${titlePart})`, async () => {
+					await supertest(admin)
+						.patch(`/${version}/device(${device1.id})`)
+						.send({ should_be_operated_by__release: getUpgradeReleaseId() })
+						.expect(200);
+					const { body } = await supertest(admin).get(
+						`/${version}/device(${device1.id})?$select=should_be_operated_by__release`,
+					);
+					expect(body.d[0]).to.not.be.undefined;
+					expect(body.d[0]['should_be_operated_by__release'].__id).to.equal(
+						getUpgradeReleaseId(),
+					);
+				});
 
-			it('should fail to downgrade', async () => {
-				await supertest(admin)
-					.patch(`/${version}/device(${device.id})`)
-					.send({ should_be_operated_by__release: prodNucHostappReleaseId })
-					.expect(400, '"Attempt to downgrade hostapp, which is not allowed"');
-			});
+				it(`should fail to downgrade (when on a release with ${titlePart})`, async () => {
+					await supertest(admin)
+						.patch(`/${version}/device(${device1.id})`)
+						.send({ should_be_operated_by__release: prodNucHostappReleaseId })
+						.expect(
+							400,
+							'"Attempt to downgrade hostapp, which is not allowed"',
+						);
+				});
 
-			it('should remove target preprovisioned hostapp, if it is an implied downgrade', async () => {
-				// if a device is preprovisioned and pinned to a release
-				// less than the version it initially checks in with, make sure the downgrade doesn't persist
-				const { body } = await supertest(admin)
-					.get(
-						`/${version}/device(${preprovisionedDevice.id})?$select=should_be_operated_by__release`,
-					)
-					.expect(200);
-				expect(body.d[0]).to.have.property(
-					'should_be_operated_by__release',
-					null,
-				);
+				it(`should remove the target preprovisioned hostapp and link with the reported upgraded ${titlePart} release, if it is an implied downgrade`, async () => {
```

Comment:
Maybe a title something like this is easier to understand: `should move device from older preprovisioned release to upgraded ${titlePart} release reported in state patch`. What do you think?